cartoframes/auth/context.py

```diff
@@ -168,24 +168,27 @@ class Context(object):
     def __init__(self, base_url=None, api_key='default_public', creds=None, session=None,
                  verbose=0):
 
-        self.creds = Credentials(creds=creds, key=api_key, base_url=base_url)
-        self.session = session
+        if creds:
+            self.creds = Credentials.from_credentials(creds)
+        else:
+            self.creds = Credentials(api_key=api_key, base_url=base_url, session=session)
+
         self.auth_client = APIKeyAuthClient(
-            base_url=self.creds.base_url(),
-            api_key=self.creds.key(),
+            base_url=self.creds.base_url,
+            api_key=self.creds.api_key,
             session=session,
             client_id='cartoframes_{}'.format(__version__),
             user_agent='cartoframes_{}'.format(__version__)
         )
         self.auth_api_client = AuthAPIClient(
-            base_url=self.creds.base_url(),
-            api_key=self.creds.key(),
+            base_url=self.creds.base_url,
+            api_key=self.creds.api_key,
             session=session
         )
         self.sql_client = SQLClient(self.auth_client)
         self.copy_client = CopySQLClient(self.auth_client)
         self.batch_sql_client = BatchSQLClient(self.auth_client)
-        self.creds.username(self.auth_client.username)
+        self.creds.username = self.auth_client.username
```

Comment:
We should use a setter in username, because we need to recompute the base_url with the new username (only if it is not set by the user)

Reply:
> We should use a setter in username

We are using it

Until now, we have not changed `base_url` when `username` is updated. I keep this comment:

```
    @property
    def username(self):
        """Credentials username

        Note:
            This does not update the `base_url` attribute. You should use `credentials.base_url = new_abse_url`
        """
        return self._username
```

Should we update `base_url` when `username` is updated and vice versa? I dont really know. I didn't want to change the behaviour because I dont know the original reason

Reply:
If we decided to change it, I will do it in the [main PR](https://github.com/CartoDB/cartoframes/pull/850)

Reply:
I originally chose to give `base_url` more preference than `username` in the constructor because it feels more specific and purposeful, so it has precedence. If you both think it makes sense to change it to be more symmetric, I support that.

Overall, I don't expect people to be using the methods on this class very much, so I doubt it will be a very big issue either option we choose.

Reply:
Ok, let's add some magic, but only for username setter

Reply:
Done in https://github.com/CartoDB/cartoframes/pull/850